Restore MySQL driver and fix ordering problem

jamescheney opened the pull request:
This PR restores the MySQL driver, a precondition for #322, and fixes #856 (albeit this could be cleaned up a bit).

Marked as WIP because the way query results are constructed by Sqlite3 is still a little inefficient and could be improved, and I'd like to make progress towards #855 also before merging, unless it turns out to be harder than I think.

- [x] Avoid unnecessary `List.rev`s in sqlite3 driver and value.ml
- [X] Basic support for INSERT RETURNING for all supported dbs.
- [x] Delete remaining dead code

jamescheney commented:
This fixes parts of #856 except for the extension to support RETURNING multiple columns which I leave as a todo.

Something horrible has happened so the PR / branch probably needs to be redone from a clean checkout of master.

SimonJF commented:
Regarding something horrible happening, have you tried:

> Here's a good workaround. Use the Edit button when viewing the PR in GitHub to change the base branch to something other than master. Then switch it back to master and it will now correctly show only the changes from the most recent commits.

(from the following link; there are also some other solutions)
 
https://stackoverflow.com/questions/16306012/github-pull-request-showing-commits-that-are-already-in-target-branch

jamescheney commented:
I seem to have found / remembered the incantation needed to get the branches synced up again, which was `git merge --no-ff master` - if that doesn't cause problems I'm inclined to leave as is.

SimonJF commented:
Ah, good to know!

jamescheney commented:
So the CI is failing because of an apparent regression on #835...

jamescheney commented:
I take that back, I only thought that because I had a stale build.  Looks like I just needed to adjust the factorials example again.

jamescheney commented:
I've decided to weaken #855 sufficiently that this fixes it.

jamescheney commented:
Now fixed sqlite3 inefficiency and realized that mysql driver suffered from a different problem (whose solution is similar).  Previously, we would randomly access the mysql library's result which was apparently slow.  Now, instead I load it all into an append-only buffer.  Possibly doing this for postgres results would be better and more uniform but this PR is already getting too big.

jamescheney commented:
The code for handling null integers in mySQL was broken, this is now fixed and I've added a database test that checks that null handling works correctly, as a first step towards #655.  This should be ready for review now.

dhil commented on core/database.ml:
```diff
@@ -91,13 +90,10 @@ let execute_insert_returning returning q db =
-                         | _ ->
-                             raise (runtime_error ("Returned the wrong number of results executing " ^ q))
-                     end
+                  if result#nfields == 1 && result#ntuples == 1
```
Structural equality `=`.

dhil commented on core/utility.ml:
```diff
@@ -1458,3 +1458,66 @@ struct
+    then begin
+      buf.nextitem <- 0;
+      buf.currpage <- buf.currpage+1;
+      if (buf.currpage == buf.numpages)
```
Structural equality `=`.

dhil commented on core/utility.ml:
```diff
@@ -1458,3 +1458,66 @@ struct
+  let set buf i x =
+  if 0 <= i && i < buf.currpage*buf.pagesize + buf.nextitem
+    then Array.set (Array.get buf.pages (i/buf.pagesize)) (i mod buf.pagesize) x
+    else raise Not_found
```
Maybe use `raise (Invalid_argument "out of bounds")`.

jamescheney replied:
OK, matching what OCaml does with ordinary arrays.

jamescheney commented on core/utility.ml:
```diff
@@ -1458,3 +1458,66 @@ struct
+    then begin 
+      buf.nextitem <- 0;
+      buf.currpage <- buf.currpage+1;
+      if (buf.currpage == buf.numpages) 
```
Done.  (But structural equality on integers behaves the same as `==`.)

jamescheney commented:
This should await #859.

jamescheney commented:
Miraculously, this seems to have commited with 859 and #861, so now that they are merged, will merge this once CI passes if no one stops me.
